features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/propagation/XRayHttpCodec.java

```diff
@@ -22,246 +22,251 @@
-        if (!isReserved(header)) {
-          additionalPart(buf, header, HttpCodec.encode(entry.getValue()), maxCapacity);
+        public Injector(Map<String, String> invertedBaggageMapping) {
+            this.invertedBaggageMapping = invertedBaggageMapping;
         }
-      }
 
-      setter.set(carrier, X_AMZN_TRACE_ID, buf.toString());
-    }
+        @Override
+        public <C> void inject(DDSpanContext context, C carrier, AgentPropagation.Setter<C> setter) {
+            long e2eStart = context.getEndToEndStartTime();
+
+            StringBuilder buf =
+                    new StringBuilder()
+                            .append(ROOT_PREFIX)
+                            .append(
+                                    String.format(
+                                            "%08x",
+                                            e2eStart > 0
+                                                    ? NANOSECONDS.toSeconds(e2eStart)
+                                                    : MILLISECONDS.toSeconds(
+                                                    context.getTrace().getTimeSource().getCurrentTimeMillis())))
+                            .append(TRACE_ID_PADDING)
+                            .append(context.getTraceId().toHexStringPadded(16))
+                            .append(';' + PARENT_PREFIX)
+                            .append(DDSpanId.toHexStringPadded(context.getSpanId()));
+
+            if (context.lockSamplingPriority()) {
+                buf.append(';' + SAMPLED_PREFIX)
+                        .append(convertSamplingPriority(context.getSamplingPriority()));
+            }
 
-    private boolean isReserved(String key) {
-      return ROOT.equals(key) || PARENT.equals(key) || SAMPLED.equals(key) || SELF.equals(key);
-    }
+            int maxCapacity = buf.length() + MAX_ADDITIONAL_BYTES;
 
-    private char convertSamplingPriority(int samplingPriority) {
-      return samplingPriority > 0 ? '1' : '0';
-    }
+            CharSequence origin = context.getOrigin();
+            if (origin != null) {
+                additionalPart(buf, ORIGIN_KEY, origin.toString(), maxCapacity);
+            }
+            if (e2eStart > 0) {
+                additionalPart(
+                        buf, E2E_START_KEY, Long.toString(NANOSECONDS.toMillis(e2eStart)), maxCapacity);
+            }
 
-    private void additionalPart(StringBuilder buf, String key, String value, int maxCapacity) {
-      if (buf.length() + key.length() + value.length() + 2 <= maxCapacity) {
-        buf.append(';').append(key).append('=').append(value);
-      }
-    }
-  }
+            for (Map.Entry<String, String> entry : context.baggageItems()) {
+                final String header;
+                if (invertedBaggageMapping.containsKey(entry.getKey())) {
+                    header = invertedBaggageMapping.get(entry.getKey());
+                } else {
+                    header = entry.getKey();
+                }
```

Comment:
can be probably a bit simpler

```suggestion
                String header = invertedBaggageMapping.get(entry.getKey());
                if (header == null) {
                    header = entry.getKey();
                }
```